### PR TITLE
babel-generator: Ensure ASCII-safe output for string literals

### LINKED ISSUE #4477: Ensure ASCII-safe output
Input:

```js
console.log(
  '\u{1D306}',
  `\u{1D306}`,
  /\u{1D306}/u
);
```

Current output: https://babeljs.io/repl/#?babili=false&evaluate=true&lineWrap=false&presets=es2015&experimental=false&loose=false&spec=false&code=console.log(%0A%20%20'%5Cu%7B1D306%7D'%2C%0A%20%20%60%5Cu%7B1D306%7D%60%2C%0A%20%20%2F%5Cu%7B1D306%7D%2Fu%0A)%3B

```js
'use strict';

console.log('𝌆', '𝌆', /(?:\uD834\uDF06)/);
```

Note that the output contains non-ASCII symbols, leading to potential encoding issues (e.g. if the server is not configured to serve the file as UTF-8) or just [confusion](https://twitter.com/_jayphelps/status/773758513667448832). By transpiling to a pair of Unicode escape sequences (i.e. `\uD834\uDF06` in this case) this issue is avoided.

cc @jayphelps 

TL;DR we should ensure any generated code for the following literals is ASCII-safe:

- [x] string literals
    - patch: #4478 
- [ ] template literals
    - patch: will submit a PR as soon as #4478 is merged
- [x] regular expression literals (taken care of by our use of [regexpu](https://mths.be/regexpu))

### PR DESCRIPTION
**Pull request by mathiasbynens**
| Q                 | A
| ----------------- | ---
| Bug fix?          | no
| Breaking change?  | no
| New feature?      | no
| Deprecations?     | no
| Spec compliancy?  | no
| Tests added/pass? | yes
| Fixed tickets     | #4477
| License           | MIT
| Doc PR            | none

Fixes #4477.

### CODE REVIEW

**Comment by mathiasbynens** [comment]
Initial approach:

```patch
From f9571f26f217ff75519291e426c8f798cb0c9a2f Mon Sep 17 00:00:00 2001
From: Mathias Bynens <mathias@qiwi.be>
Date: Thu, 8 Sep 2016 08:04:20 +0200
Subject: [PATCH] transform-es2015-literals: Ensure ASCII-safe output

Fixes #4477.
---
 .../babel-plugin-transform-es2015-literals/package.json  |  3 ++-
 .../babel-plugin-transform-es2015-literals/src/index.js  | 16 ++++++++++++++--
 .../fixtures/unicode-code-point-escapes/basic/actual.js  |  2 ++
 .../unicode-code-point-escapes/basic/expected.js         |  2 ++
 .../fixtures/unicode-code-point-escapes/options.json     |  3 +++
 .../babel-plugin-transform-es2015-literals/test/index.js |  1 +
 6 files changed, 24 insertions(+), 3 deletions(-)
 create mode 100644 packages/babel-plugin-transform-es2015-literals/test/fixtures/unicode-code-point-escapes/basic/actual.js
 create mode 100644 packages/babel-plugin-transform-es2015-literals/test/fixtures/unicode-code-point-escapes/basic/expected.js
 create mode 100644 packages/babel-plugin-transform-es2015-literals/test/fixtures/unicode-code-point-escapes/options.json
 create mode 100644 packages/babel-plugin-transform-es2015-literals/test/index.js

diff --git a/packages/babel-plugin-transform-es2015-literals/package.json b/packages/babel-plugin-transform-es2015-literals/package.json
index 0d3047c..da86212 100644
--- a/packages/babel-plugin-transform-es2015-literals/package.json
+++ b/packages/babel-plugin-transform-es2015-literals/package.json
@@ -9,7 +9,8 @@
     "babel-plugin"
   ],
   "dependencies": {
-    "babel-runtime": "^6.0.0"
+    "babel-runtime": "^6.0.0",
+    "jsesc": "^1.3.0"
   },
   "devDependencies": {
     "babel-helper-plugin-test-runner": "^6.8.0"
diff --git a/packages/babel-plugin-transform-es2015-literals/src/index.js b/packages/babel-plugin-transform-es2015-literals/src/index.js
index 3f6d33c..ef47cfb 100644
--- a/packages/babel-plugin-transform-es2015-literals/src/index.js
+++ b/packages/babel-plugin-transform-es2015-literals/src/index.js
@@ -1,17 +1,29 @@
+import jsesc from "jsesc";
+
+const regexPrintableAscii = /[\x20-\x7E]/;
+
 export default function () {
   return {
     visitor: {
       NumericLiteral({ node }) {
-        // number octal like 0b10 or 0o70
+        // Numeric octal literal like 0b10 or 0o70
         if (node.extra && /^0[ob]/i.test(node.extra.raw)) {
           node.extra = undefined;
         }
       },
 
       StringLiteral({ node }) {
-        // unicode escape
+        // Unicode code point escape like \u{1D306}
         if (node.extra && /\\[u]/gi.test(node.extra.raw)) {
           node.extra = undefined;
+          const symbol = node.value;
+          if (!regexPrintableAscii.test(symbol)) {
+            node.extra = {
+              raw: jsesc(symbol, {
+                wrap: true
+              })
+            };
+          }
         }
       }
     }
diff --git a/packages/babel-plugin-transform-es2015-literals/test/fixtures/unicode-code-point-escapes/basic/actual.js b/packages/babel-plugin-transform-es2015-literals/test/fixtures/unicode-code-point-escapes/basic/actual.js
new file mode 100644
index 0000000..d528df1
--- /dev/null
+++ b/packages/babel-plugin-transform-es2015-literals/test/fixtures/unicode-code-point-escapes/basic/actual.js
@@ -0,0 +1,2 @@
+var string = "foo\u{1D306}bar\u{A}baz\u{0000000000}qux";
+var template = `foo\u{1D306}bar\u{A}baz\u{0000000000}qux`;
diff --git a/packages/babel-plugin-transform-es2015-literals/test/fixtures/unicode-code-point-escapes/basic/expected.js b/packages/babel-plugin-transform-es2015-literals/test/fixtures/unicode-code-point-escapes/basic/expected.js
new file mode 100644
index 0000000..f3828a4
--- /dev/null
+++ b/packages/babel-plugin-transform-es2015-literals/test/fixtures/unicode-code-point-escapes/basic/expected.js
@@ -0,0 +1,2 @@
+var string = "foo\uD834\uDF06bar\nbaz\0qux";
+var template = "foo\uD834\uDF06bar\nbaz\0qux";
diff --git a/packages/babel-plugin-transform-es2015-literals/test/fixtures/unicode-code-point-escapes/options.json b/packages/babel-plugin-transform-es2015-literals/test/fixtures/unicode-code-point-escapes/options.json
new file mode 100644
index 0000000..4093e61
--- /dev/null
+++ b/packages/babel-plugin-transform-es2015-literals/test/fixtures/unicode-code-point-escapes/options.json
@@ -0,0 +1,3 @@
+{
+  "plugins": ["transform-es2015-literals"]
+}
diff --git a/packages/babel-plugin-transform-es2015-literals/test/index.js b/packages/babel-plugin-transform-es2015-literals/test/index.js
new file mode 100644
index 0000000..1f6634a
--- /dev/null
+++ b/packages/babel-plugin-transform-es2015-literals/test/index.js
@@ -0,0 +1 @@
+require("babel-helper-plugin-test-runner")(__dirname);
```

However, it seems better to handle this in babel-generator itself. https://github.com/babel/babel/blob/2b4e49dd6496945e3e490d84b2a635cf730d15e2/packages/babel-generator/src/generators/types.js#L138-L143

File: packages/babel-generator/test/fixtures/minified/literals/actual.js [hunk 1]
```diff
@@ -4,6 +4,7 @@
 '\x20';
 "\n\r";
 "😂";
+`😂`;
```
Comment (mathiasbynens) [inline packages/babel-generator/test/fixtures/minified/literals/actual.js]: I’ve added this test to indicate that this patch only handles string literals and not template literals. (Regular expression literals are already taken care of through our use of [regexpu](https://mths.be/regexpu).)

File: packages/babel-generator/src/generators/types.js [hunk 2]
```diff
@@ -135,26 +136,11 @@ export function StringLiteral(node: Object, parent: Object) {
...
-  let val = JSON.stringify(node.value);
-
-  // escape illegal js but valid json unicode characters
-  val = val.replace(/[\u000A\u000D\u2028\u2029]/g, function (c) {
-    return "\\u" + ("0000" + c.charCodeAt(0).toString(16)).slice(-4);
+  // ensure the output is ASCII-safe
+  let val = jsesc(node.value, {
+    quotes: t.isJSX(parent) ? "double" : this.format.quotes,
+    wrap: true
   });
```
Comment (mathiasbynens) [inline packages/babel-generator/src/generators/types.js]: The use of jsesc also allows us to get rid of the hacky parts below (i.e. remove double quotes, unescape double quotes, escape single quotes, add single quotes).

File: packages/babel-generator/package.json [hunk 3]
```diff
@@ -15,6 +15,7 @@
     "babel-runtime": "^6.9.0",
     "babel-types": "^6.14.0",
     "detect-indent": "^3.0.1",
+    "jsesc": "^1.3.0",
```
Comment (loganfsmyth) [inline packages/babel-generator/package.json]: Looks like this is Node >= 4 so I don't think we can land this at the moment. We all agree that we want to bump to >= 4 AFAIK, but we won't be able to do it until we bump our major version, which we haven't really had the opportunity to plan yet.

Comment (danez) [inline packages/babel-generator/package.json]: AFAIK it looks to me that 1.x supports versions < 4 and 2.0 only supports 4+
https://github.com/mathiasbynens/jsesc/commit/e212a67946cd2a31578fac650f31281e52e4ca0c
But haven't tested.

Comment (loganfsmyth) [inline packages/babel-generator/package.json]: Ahh my mistake then, didn't compare major versions.

Comment (mathiasbynens) [inline packages/babel-generator/package.json]: @danez is right. I’ve used jsesc@1.3.0 because it supports older versions of Node.js.

Comment (loganfsmyth) [inline packages/babel-generator/package.json]: Thanks!

Comment (hzoo) [inline packages/babel-generator/package.json]: Also CI was passing on node 0.10 😄

**Comment by hzoo** [comment]
Acutally, it would be nice if the changelog included new/removed deps

File: packages/babel-generator/src/generators/types.js (hunk 2 above)
Comment (hzoo) [inline packages/babel-generator/src/generators/types.js]: Nice to remove all that 😄

Comment (kangax) [inline packages/babel-generator/src/generators/types.js]: Fun fact: I updated Babel to 6.17 at Facebook and it broke one of our (PHP) parsers and the entire transform pipeline as a result :D After tedious investigation, I think this is the culprit. When someone uses "&times;" in our codebase (within custom tags) it's then transformed to '×' which Babel represented as '\u00D7' before. Now `jsesc` turns it into '\xD7'. Apparently PHP parser doesn't understand the latter. ¯\_(ツ)_/¯

Comment (mathiasbynens) [inline packages/babel-generator/src/generators/types.js]: @kangax Which PHP parser is that?

Comment (kangax) [inline packages/babel-generator/src/generators/types.js]: Dear god... Tracked it down further to `json_decode` returning `null` whenever there's \x in a string.

**Pull request merged**